### ISSUE TITLE
Report `\0` for `no-control-regex`

### ISSUE THREAD

**Issue by brettz9** [eslint/eslint#7476]
**Tell us about your environment**
- **ESLint Version:** 3.9.0 or http://eslint.org/demo/
- **Node Version:** 4.6.1
- **npm Version:** 3.10.9

**What parser (default, Babel-ESLint, etc.) are you using?** Default

**Please show your full configuration:**

```
no-control-regex: true
```

**What did you do? Please include the actual source code causing the issue.**

``` js
var regex = /\0/;
```

**What did you expect to happen?**

Get flagged as a control regex.

**What actually happened? Please include the actual, raw output from ESLint.**

No warnings or errors.

Note that you recently fixed such `\0` control characters not to get flagged as useless escapes,  but this warning should presumably still be triggered.

**Comment by not-an-aardvark** [comment]
Hi, thanks for the report.

I think this is a duplicate of https://github.com/eslint/eslint/issues/7472, so I'm going to close this issue. Please follow that issue for updates.

**Comment by not-an-aardvark** [comment]
Actually wait, I misread this issue. Reopening, sorry about that.

**Comment by not-an-aardvark** [comment]
I noticed that you'received using `no-control-regex: true`, which isn'the a valid configuration as far as I know. What happens if you change it to `no-control-regex: "error"` instead?

**Comment by brettz9** [comment]
Sorry, I was just representing the demo checkbox... I am using eslint-config-standard which has `"no-control-regex": 2`.

**Comment by not-an-aardvark** [comment]
Thanks, I can reproduce this.

Note to the rest of the team: This is *not* a regression (the bug also exists in previous versions).

**Comment by brettz9** [comment]
Oh, and I also see that `\cX` (where X is A-Z) isn't flagged either though it does produce a control character also. And maybe `\f`, `\v`, and `[\b]` are rare enough control characters to deserve being flagged.

Of course `\n` and `\r` are certainly frequent enough not to be included (and I don't think using a Unicode escape sequence representing a control character like `\u0000` or `\u{0}` ought to be flagged). It'd probably be worth mentioning in the docs though which such escapes are indeed deliberately not going to be marked.

**Comment by not-an-aardvark** [comment]
Good point about `\n` and `\r`. Would you be interested in making a PR to mention that in the docs?

**Comment by brettz9** [comment]
I can see about a PR (including code) which:

1. prohibits `\0`, `\cX`, `\f`, `\v`, and `[\b]`
2. mentions `\n`, `\r`, and `\u0000` or `\u{0}`-style escapes of Unicode control characters (i.e., for `[\u0000-\u001F\u007F\u0080-\u009F]`) as specifically allowable

Just want to make sure this is the desired way forward.

**Comment by not-an-aardvark** [comment]
Hmm, I'm reconsidering this after looking at the rule again.

It looks like we disallow things like `\x10`, even though we allow things like `\n`. I think the point of the rule is to prevent the use of control characters with syntax like `\x00` or `\u0000`, where the use of a control character might be accidental. With `\0`, `\cX`, `\f`, `\v`, and `[\b]`, I think it's more likely that the use of a control character was intentional, so maybe we shouldn't report those cases (for the same reason that we report `\x10` and not `\n`)

However, I think `\uXXXX` and `\u{XXXXX}` are similar to `\x00` in that they're specifying control *codes* which might accidentally correspond to control characters rather than other unicode characters. So we should report these cases.

In other words, I think maybe the rule is more about disallowing the use of control characters via control code, rather than disallowing escapes that happen to correspond to control characters.

**Comment by brettz9** [comment]
Makes sense...

(In that case, `\uXXXX` is probably even more significant than `\x00` since the proportion of `\x00` that are control characters are greater.)

Let me know what the decision is.

**Comment by brettz9** [comment]
OTOH, there's a case to be made that having a uniform policy to require use of the more expressive though more verbose Unicode escape sequences could prevent errors along the lines of accidental backslashes.

Unlike for prohibiting `\uXXXX`, prohibiting the more obscure `\v`, `\cX`, etc. allows for a consistent policy whereby one can continue to occasionally use control characters without needing to disable the rule.

**Comment by not-an-aardvark** [comment]
Closing because I think the current behavior (only reporting unicode escapes like `\uXXXX` is working as intended.
